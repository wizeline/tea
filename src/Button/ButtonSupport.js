// @flow

import { css } from 'styled-components';
import type { ButtonProps } from './ButtonTypes';

const circleSizeSuport = (small, big) => {
  const baseCSS = css`
    border-radius: 50%;
    box-shadow: unset;
    min-height: unset;
    min-width: unset;
    padding: 0;

    & > .svgContainer {
      align-items: center;
      justify-content: center;
    }
  `;
  if (small) {
    return css`
      ${baseCSS};
      height: 0.75rem;
      width: 0.75rem;

      & > .svgContainer {
        height: 0.375rem;
        width: 0.375rem;
      }
    `;
  } else if (big) {
    return css`
      ${baseCSS};
      height: 2.25rem;
      width: 2.25rem;

      & > .svgContainer {
        height: 1rem;
        width: 1rem;
      }
    `;
  }
  return css`
    ${baseCSS};
    height: 1.25rem;
    width: 1.25rem;

    & > .svgContainer {
      height: 0.5rem;
      width: 0.5rem;
    }
  `;
};

const buttonSizeSupport = ({ small, big, circle }: ButtonProps) => {
  const circleStyles = circle ? circleSizeSuport(small, big) : null;
  if (small) {
    return css`
      border-radius: 0.1875rem;
      box-shadow: 0 0.125rem 0.0625rem 0 rgba(0, 0, 0, 0.03);
      font-size: 0.75rem;
      min-height: 1.4rem;
      min-width: 4.5rem;
      padding: 0 1rem;
      ${circleStyles};
    `;
  }
  if (big) {
    return css`
      border-radius: 0.3125rem;
      box-shadow: 0 0.125rem 0.0625rem 0 rgba(0, 0, 0, 0.03),
        inset 0 -0.1875rem 0 0 rgba(0, 0, 0, 0.08);
      font-size: 1rem;
      min-height: 3rem;
      min-width: 10.25rem;
      padding: 0 1.5rem;
      ${circleStyles};
    `;
  }
  return css`
    border-radius: 0.1875rem;
    box-shadow: 0 0.125rem 0.0625rem 0 rgba(0, 0, 0, 0.03);
    font-size: 0.875rem;
    min-height: 2rem;
    min-width: 6.25rem;
    padding: 0 1rem;
    ${circleStyles};
  `;
};

const iconButtonSupport = ({ circle, icon, children }: ButtonProps) => {
  if (!circle && icon && !children[1]) {
    return css`
      min-width: 3rem;
    `;
  }
  if (!circle && icon && children[1]) {
    return css`
      & > .svgContainer {
        height: 1rem;
        margin-right: 0.5rem;
        width: 1rem;
      }
    `;
  }
  return null;
};

export { buttonSizeSupport, iconButtonSupport };
