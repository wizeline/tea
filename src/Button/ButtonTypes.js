// @flow
import type { Node } from 'react';

export type ButtonProps = {
  // Color
  destructive?: boolean,
  primary?: boolean,
  secondary?: boolean,
  // Size
  big?: boolean,
  normal?: boolean,
  small?: boolean,
  // Other types
  disabled?: boolean,
  icon?: string,
  circle?: boolean,
  text?: string,
  children: [Node, Node],
};
