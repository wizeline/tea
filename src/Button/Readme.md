```jsx
<div>
  <h1>Primary</h1>
  <table>
    <thead>
      <tr>
        <td>Small</td>
        <td>Normal</td>
        <td>Big</td>
        <td>With Icon</td>
        <td>Icon-only</td>
        <td>Disabled</td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          <Button small text="Create" />
          <Button small round icon="dots-h" text="Create" />
        </td>
        <td>
          <Button text="Create" />
          <Button round icon="dots-h" text="Create" />
        </td>
        <td>
          <Button big text="Create" />
          <Button big round icon="dots-h" text="Create" />
        </td>
        <td>
          <Button icon="plus" text="Create" />
        </td>
        <td>
          <Button icon="plus" />
        </td>
        <td>
          <Button disabled icon="plus" text="Create" />
        </td>
      </tr>
    </tbody>
  </table>
  <h1>Secondary</h1>
  <table>
    <tbody>
      <tr>
        <td>
          <Button secondary small text="Create" />
          <Button secondary small round icon="dots-h" text="Create" />
        </td>
        <td>
          <Button secondary text="Create" />
          <Button secondary round icon="dots-h" text="Create" />
        </td>
        <td>
          <Button secondary big text="Create" />
          <Button secondary big round icon="dots-h" text="Create" />
        </td>
        <td>
          <Button secondary icon="plus" text="Create" />
        </td>
        <td>
          <Button secondary icon="plus" />
        </td>
        <td>
          <Button disabled secondary icon="plus" text="Create" />
        </td>
      </tr>
    </tbody>
  </table>
  <h1>Destructive</h1>
  <table>
    <tbody>
      <tr>
        <td>
          <Button destructive small text="Create" />
          <Button destructive small round icon="dots-h" text="Create" />
        </td>
        <td>
          <Button destructive text="Create" />
          <Button destructive round icon="dots-h" text="Create" />
        </td>
        <td>
          <Button destructive big text="Create" />
          <Button destructive big round icon="dots-h" text="Create" />
        </td>
        <td>
          <Button destructive icon="plus" text="Create" />
        </td>
        <td>
          <Button destructive icon="plus" />
        </td>
        <td>
          <Button disabled destructive icon="plus" text="Create" />
        </td>
      </tr>
    </tbody>
  </table>
</div>
```
