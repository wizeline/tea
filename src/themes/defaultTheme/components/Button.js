// @flow

import defaultThemeBackground from '../../styles/backgrounds';
import defaultThemeBorder from '../../styles/borders';
import defaultThemeText from './Text';

export default {
  ...defaultThemeBackground,
  ...defaultThemeBorder,
  ...defaultThemeText,
};
