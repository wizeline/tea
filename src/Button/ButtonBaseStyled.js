// @flow
/* eslint flowtype-errors/enforce-min-coverage: 0 */

import styled from 'styled-components';
import { buttonSizeSupport, iconButtonSupport } from './ButtonSupport';
import { themeProp } from '../utils';
import defaultButtonTheme from '../themes/defaultTheme/components/Button';

const BaseButtonStyled = styled.button`
  display: flex;
  align-items: center;
  justify-content: center;
  transition: all 0.1s;
  ${buttonSizeSupport};
  ${iconButtonSupport};

  & > .svgContainer {
    display: flex;
  }
`;

const DarkerButton = BaseButtonStyled.extend`
  color: ${themeProp('textPrimaryInvert')};

  &:disabled,
  &:disabled:hover,
  &:disabled:active {
    background-color: ${themeProp('backgroundPrimaryDisabled')};
    border: ${themeProp('backgroundPrimaryDisabled')};
    box-shadow: none;
    color: ${themeProp('textPrimaryDisabled')};

    & > .svgContainer {
      svg {
        fill: ${themeProp('textPrimaryDisabled')};
      }
    }
  }
`;

const LighterButton = BaseButtonStyled.extend`
  color: ${themeProp('textPrimary')};

  &:disabled,
  &:disabled:hover,
  &:disabled:active {
    border: solid 1px ${themeProp('borderSecondaryDisabled')};
    color: ${themeProp('textSecondaryDisabled')};

    & > .svgContainer {
      svg {
        fill: ${themeProp('textSecondaryDisabled')};
      }
    }
  }
`;

BaseButtonStyled.defaultProps = {
  theme: defaultButtonTheme,
};

export { DarkerButton, LighterButton };
