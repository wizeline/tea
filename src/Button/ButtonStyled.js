// @flow

import { lighten } from 'polished';
import compose from 'ramda/src/compose';
import { DarkerButton, LighterButton } from './ButtonBaseStyled';
import { themeProp } from '../utils';
import themeConstants from '../themes/defaultTheme/constants';

const {
  BACKGROUND_DANGER,
  BACKGROUND_PRIMARY,
  BACKGROUND_SECONDARY_HOVER,
  BACKGROUND_SECONDARY,
  BORDER_SECONDARY_HOVER,
  BORDER_SECONDARY,
  LIGHTEN_BACKGROUND_PERCENTAGE,
} = themeConstants;

const getHoverColor = (propName: string) =>
  compose(lighten(LIGHTEN_BACKGROUND_PERCENTAGE), themeProp(propName));

const getDarkerButtonVariant = baseName => DarkerButton.extend`
  background-color: ${themeProp(baseName)};
  border: ${themeProp(baseName)};

  &:hover,
  &:active {
    background-color: ${getHoverColor(baseName)};
    border: ${themeProp(baseName)};
  }
`;

const PrimaryButton = getDarkerButtonVariant(BACKGROUND_PRIMARY);

const DestructiveButton = getDarkerButtonVariant(BACKGROUND_DANGER);

const SecondaryButton = LighterButton.extend`
  background-color: ${themeProp(BACKGROUND_SECONDARY)};
  border: solid 1px ${themeProp(BORDER_SECONDARY)};
  box-shadow: 0 0.125rem 0.0625rem 0 rgba(0, 0, 0, 0.03);

  &:hover,
  &:active {
    background-color: ${themeProp(BACKGROUND_SECONDARY_HOVER)};
    border: solid 1px ${themeProp(BORDER_SECONDARY_HOVER)};
  }
`;

export { PrimaryButton, DestructiveButton, SecondaryButton };
