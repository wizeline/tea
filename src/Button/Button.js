// @flow

import React from 'react';
import ButtonStyled from './ButtonStyled';
import type { ButtonProps } from './ButtonTypes';

const Button = (props: ButtonProps) => (
  <ButtonStyled>{props.text}</ButtonStyled>
);

export default Button;
