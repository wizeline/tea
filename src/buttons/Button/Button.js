// @flow

import React from 'react';
import { ButtonStyled, TextWrapperStyled } from './ButtonStyled';
import type { ButtonPropTypes } from './ButtonTypes';
import SVGIcon from '../../icons/SVGIcon';

const propGroups = {
  priority: ['primary', 'secondary', 'danger'],
  size: ['small', 'medium', 'large'],
  defaultPriority: 'primary',
  defaultSize: 'medium',
};

const supportPropGroup = (
  groupName: string,
  propGroup: Array<string>,
  inProps: Object,
  defaultProp: string,
) => {
  let truthyProp = null;
  propGroup.some(prop => {
    if (inProps[prop]) {
      truthyProp = prop;
      return true;
    }
    return false;
  });
  return { [groupName]: truthyProp || defaultProp };
};

const Button = (props: ButtonPropTypes) => {
  const { text, iconName, disabled, round } = props;
  const iconComponent = iconName ? <SVGIcon name={iconName} /> : null;
  const buttonStyledProps = {
    ...supportPropGroup(
      'priority',
      propGroups.priority,
      props,
      propGroups.defaultPriority,
    ),
    ...supportPropGroup('size', propGroups.size, props, propGroups.defaultSize),
    disabled,
    round,
  };

  const labelComponent =
    text && !round ? <TextWrapperStyled>{text}</TextWrapperStyled> : null;
  return (
    <ButtonStyled {...buttonStyledProps}>
      {iconComponent}
      {labelComponent}
    </ButtonStyled>
  );
};

export default Button;
