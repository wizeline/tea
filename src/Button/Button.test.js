import React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';

describe('Button', () => {
  it('renders correctly when given icon diabled and text props', () => {
    const wrapper = shallow(<Button icon="dots-h" disabled text="Test Text" />);
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly when given icon and no text props', () => {
    const wrapper = shallow(<Button icon="dots-h" />);
    expect(wrapper).toHaveStyleRule('min-width', '48px');
  });

  it('renders correctly when given destructive prop', () => {
    const wrapper = shallow(<Button destructive text="Test Text" />);
    expect(wrapper).toHaveStyleRule('background-color', '#ec5659');
    expect(wrapper).toHaveStyleRule('border', '#ec5659');
  });

  it('renders correctly when given secondary prop', () => {
    const wrapper = shallow(<Button secondary text="Test Text" />);
    expect(wrapper).toHaveStyleRule('background-color', '#ffffff');
    expect(wrapper).toHaveStyleRule('border', 'solid 1px #a2abaf');
  });

  it('renders correctly when given big prop', () => {
    const wrapper = shallow(<Button big text="Test Text" />);
    expect(wrapper).toHaveStyleRule('font-size', '1rem');
    expect(wrapper).toHaveStyleRule('min-height', '3rem');
    expect(wrapper).toHaveStyleRule('min-width', '10.25rem');
  });

  it('renders correctly when given small prop', () => {
    const wrapper = shallow(<Button small text="Test Text" />);
    expect(wrapper).toHaveStyleRule('font-size', '0.75rem');
    expect(wrapper).toHaveStyleRule('min-height', '1.4rem');
    expect(wrapper).toHaveStyleRule('min-width', '4.5rem');
  });

  it('renders correctly when given big and round props', () => {
    const wrapper = shallow(<Button big round text="Test Text" />);
    expect(wrapper).toHaveStyleRule('font-size', '1rem');
    expect(wrapper).toHaveStyleRule('height', '2.25rem');
    expect(wrapper).toHaveStyleRule('width', '2.25rem');
  });

  it('renders correctly when given small and round props', () => {
    const wrapper = shallow(<Button small round text="Test Text" />);
    expect(wrapper).toHaveStyleRule('font-size', '0.75rem');
    expect(wrapper).toHaveStyleRule('height', '0.75rem');
    expect(wrapper).toHaveStyleRule('width', '0.75rem');
  });

  it('renders correctly with round prop', () => {
    const wrapper = shallow(<Button round />);
    expect(wrapper).toHaveStyleRule('border-radius', '50%');
  });
});
