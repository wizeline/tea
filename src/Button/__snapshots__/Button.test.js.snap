// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button renders correctly when given icon diabled and text props 1`] = `
<styled.button
  disabled={true}
  icon="dots-h"
  text="Test Text"
  theme={
    Object {
      "backgroundDanger": "#ED524C",
      "backgroundDangerDisabled": "#E6E7E8",
      "backgroundInput": "#FFFFFF",
      "backgroundPrimary": "#50A5E7",
      "backgroundPrimaryDisabled": "#E6E7E8",
      "backgroundSecondary": "#FFFFFF",
      "backgroundSecondaryDisabled": "#FFFFFF",
      "backgroundSecondaryHover": "#FFFFFF",
      "borderInput": "#E6E7E8",
      "borderInputFocus": "#50A5E7",
      "borderInputInvalid": "#ED524C",
      "borderSecondary": "#8B9193",
      "borderSecondaryDisabled": "#E6E7E8",
      "borderSecondaryHover": "#50A5E7",
      "textBody": "#3F484B",
      "textDanger": "#FFFFFF",
      "textDangerDisabled": "#CDD0D2",
      "textDangerHover": "#FFFFFF",
      "textHeading": "#3F484B",
      "textIcon": "#FFFFFF",
      "textIconInvert": "#222729",
      "textInput": "#3F484B",
      "textInputFocus": "#8B9193",
      "textInputInvalid": "#E92720",
      "textInputPlaceholder": "#8B9193",
      "textPrimary": "#3F484B",
      "textPrimaryDisabled": "#8B9193",
      "textPrimaryHover": "#FFFFFF",
      "textPrimaryInvert": "#FFFFFF",
      "textSecondary": "#3F484B",
      "textSecondaryDisabled": "#CDD0D2",
      "textSecondaryHover": "#FFFFFF",
      "textSubtitle": "#8B9193",
    }
  }
>
  <div
    className="svgContainer"
  >
    <SVGIcon
      invert={true}
      name="dots-h"
      theme={
        Object {
          "backgroundDanger": "#ED524C",
          "backgroundDangerDisabled": "#E6E7E8",
          "backgroundInput": "#FFFFFF",
          "backgroundPrimary": "#50A5E7",
          "backgroundPrimaryDisabled": "#E6E7E8",
          "backgroundSecondary": "#FFFFFF",
          "backgroundSecondaryDisabled": "#FFFFFF",
          "backgroundSecondaryHover": "#FFFFFF",
          "textBody": "#3F484B",
          "textDanger": "#FFFFFF",
          "textDangerDisabled": "#CDD0D2",
          "textDangerHover": "#FFFFFF",
          "textHeading": "#3F484B",
          "textIcon": "#FFFFFF",
          "textIconInvert": "#222729",
          "textInput": "#3F484B",
          "textInputFocus": "#8B9193",
          "textInputInvalid": "#E92720",
          "textInputPlaceholder": "#8B9193",
          "textPrimary": "#3F484B",
          "textPrimaryDisabled": "#8B9193",
          "textPrimaryHover": "#FFFFFF",
          "textPrimaryInvert": "#FFFFFF",
          "textSecondary": "#3F484B",
          "textSecondaryDisabled": "#CDD0D2",
          "textSecondaryHover": "#FFFFFF",
          "textSubtitle": "#8B9193",
        }
      }
    />
  </div>
  Test Text
</styled.button>
`;
